man/man1/regridclm.1
.TH regridclm 1  "USER COMMANDS"
.SH NAME
regridclm \- regrid CLM file to new grid
.SH SYNOPSIS
.B regridclm
[\-h] [\-v] [\-metafile] [\-size4] [\-byte] [\-search] [\-zero] [\-longheader] [\-json]
.I [grid_old.clm] grid_new.clm data_old.clm data_new.clm
.SH DESCRIPTION
Program reads CLM file and corresponding grid file to regrid the data to a new grid file.
.SH OPTIONS
.TP
\-h,\--help
Print a short help text.
.TP
\-v,--version
Print version.
.TP
\-metafile
Input files are JSON metafiles describing the structure of the data. Option implicitely sets \-json option and an additional JSON metafile with suffix \fI.json\fP is written for output file.
.TP
\-size4
Size of CLM input data set to 4 bytes. By default size is read from header or set to 2 bytes for version<3.
.TP
\-byte
Size of CLM input data set to one byte. By default size is read from header or set to 2 bytes for version<3.
.TP
\-search
If grid coordinate cannot be found in old grid nearest grid cell is used.
.TP
\-zero
If grid coordinate cannot be found in old grid values are set to zero.
.TP
\-longheader
Version of CLM file is set to 2. By default version is read from file header.
.TP
\-json
An additional JSON metafile with suffix \fI.json\fP is written..
.TP
.I grid_old.clm
Corresponding grid filename of input data. Can be omitted if -metafile option is set and name of grid file can be read from JSON metafile.
.TP
.I grid_new.clm
New grid filename data should be regridded to.
.TP
.I  data_old.clm
Filename of data file that should be regridded.
.TP
.I data_new.clm
Filename of data file where regridded data is written.
.SH EXAMPLES
.TP
Regrid CLM file to the grid file of the amazon basin:
.B regridclm
grid.clm grid_amazon.clm temp_1901-2018.clm temp_amazon_1901-2018.clm
.TP
Regrid CLM file to the grid file of the amazon basin using JSON metafiles:
.B regridclm
-metafile grid_amazon.clm temp_1901-2018.clm.json temp_amazon_1901-2018.clm

.SH EXIT STATUS
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
regridlpj(1), regridsoil(1), lpjml(1), txt2clm(1), cru2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security


## [Unreleased]


## [6.1.4] - 2026-07-03

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Fabian Stenzel (stenzel@pik-potsdam.de), David Hötten (davidho@pik-potsdam.de)

### Added

- Utility `cmprestart` added to compare two restart files and check for identity.
- Function `bstruct_skiparray()` and corresponding test `test_skiparray.c` added in `bstruct` library to skip arrary elements.
- Function `bstruct_gettype()` added to `bstruct` library to get datatype of object in restart file.
- Function `bstrucht_printnamestack()` added for more comprehensive error messages.
- Comparison of restart files with `cmprestart` added to merge request template.

### Changed

- Index array object is always skipped in `bstruct_findobject.c`, test `test_indexarray.c` updated to test for reading array without seeking to an index array.
- Check for seeking to position within file added to `bstruct_findobject.c`, `bstruct_open.c`, `bstruct_wopen.c` and `bstruct_seekindexarray.c`.
- Constant `MAXLEVEL` renamed to `BSTRUCT_MAXLEVEL`.
- Macros `getname()` and `isinvalidtoken()` renamed to `bstruct_getname()` and `bstruct_isinvalidtoken()`.
- String comparison in `bstruct_fprintdata` replaced by call to `isnan()` and `isinf()`.
- New token `bstruct_dzero` defined for restart files to distinguish between float and double zeros, version of bstruct files changed to 2.


## [6.1.3] - 2026-07-03

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth(wirth@pik-potsdam.de)
- code review: Sebastian Ostberg (ostberg@pik-potsdam.de), Jannes Breier (breier@pik-potsdam.de)

### Added

- Option `-metafile` and `-json` added to `cutclm` and `catclm`.
- Option `-double` added to `cdf2soil.c` to write grid file with double datatype.

### Changed

- New datatype `Metadata` defined in `types.h` combining maps, attributes, basetemp and hlimit arrays used by `opendata()`.
- basetemp, countrymap and hlimit arrays are copied from JSON metafiles to newly created metafiles in `regridclm`, `cutlcm`, `catclm`, and `mathclm`.
- Scaling to sum of stand fractions removed for output `"vegc_avg"`.
- Utility `getcountry` reads now JSON metafile for country codes including the alpha-3 country codes and global grid instead of CLM file.
- If `-pedantic` option is set, non-matching titles in input files stops `lpjml` with an error.
- Datatype of `lon`, `lat`, and `area` elements of `Coord` changed from `Real` to `double`.
- Missing path to grid file added in `openmetafile.c`.

### Fixed

- Warning for float coordinates corrected in `cdf2coord.c`, `cdf2grid.c` and `cdf2soil.c`.
- Variables initialized to zero in `reservoir2cdf.c` and `restart2yaml.c` to avoid warnings by `gcc`.
- Carbon balance fixed in `reclaim_land.c`.
- Check for NULL pointer added in `getuser.c` to avoid SEGV if user name cannot be resolved.
- Verbose output corrected for input without time axis in `cdf2clm.c`.
- Output is rescaled by the actual sum of stand fractions in `update_daily_cell.c` and `fwriteoutput.c` to avoid division by zero (issue #411).
- Default timestep set to 1 in `regridclm.c`.
- Variable `depth` initialized to zero before layer loop in `fwriteoutput.c` to correct `soilc_1m` output.
- Length of buffer for hostname corrected in `gethost.c`.
- Output of unprintable characters corrected in `fputprintable.c`.
- Memory leak fixed in `openmetafile.c` in case of error.


## [6.1.2] - 2026-06-16

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Fixed

- Check for last year for lightning and human ignition disabled in `filesexist.c`, `lpjcheck` does not return an error for standard input.


## [6.1.1] - 2026-06-12

### Contributors

author: Jannes Breier (breier@pik-potsdam.de),  Werner von Bloh (bloh@pik-potsdam.de)

src/utils/regridclm.c
/**************************************************************************************/
/**                                                                                \n**/
/**                    r  e  g  r  i  d  c  l  m  .  c                             \n**/
/**                                                                                \n**/
/**     CLM data is regridded to new grid file                                     \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define USAGE "Usage: %s [-h] [-v] [-metafile] [-size4] [-byte] [-search] [-zero] [-longheader] [-json] [grid_old.clm] grid_new.clm data_old.clm data_new.clm\n"
#define ERR_USAGE USAGE "\nTry \"%s --help\" for more information.\n"

int main(int argc,char **argv)
{
  FILE *file,*data_file;
  Header header,header2;
  Coord *c,*c2;
  double dist_min;
  Bool swap,isint,issearch,iszero,isjson,ismeta,isbyte;
  short *data=NULL;
  int *zero=NULL;
  int *idata=NULL;
  Byte *bdata=NULL;
  long long size;
  Coord res,res2;
  Coordfile grid;
  int i,j,*index,data_version,setversion,ngrid,ngrid2,iarg,format;
  int index_datafile,index_gridfile,rc;
  float lon,lat,*fzero;
  char *arglist,*out_json;
  const char *progname;
  size_t offset;
  String id;
  Filename filename,grid_name;
  Type grid_type;
  Metadata metadata;
  isint=issearch=iszero=isjson=ismeta=isbyte=FALSE;
  setversion=READ_VERSION;
  grid_name.name=NULL;
  initmetadata(&metadata,NULL);
  progname=strippath(argv[0]);
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-h") || !strcmp(argv[iarg],"--help"))
      {
        printf("     ");
        rc=printf("%s (" __DATE__ ") Help",progname);
        printf("\n     ");
        repeatch('=',rc);
        printf("\n\nRegrid CLM file to new grid for LPJmL version %s\n",getversion());
        printf(USAGE
               "\nArguments:\n"
               "-h,--help    print this help text\n"
               "-v,--version print LPJmL version\n"
               "-metafile    input files are JSON metafiles,\n"
               "             additional JSON file written for outputs"
               "-size4       size of CLM input data is set to 4 bytes\n"
               "-byte        size of CLM input data is set to 1 byte\n"
               "-search      if cell not found, nearest cell is used\n"
               "-zero        if cell not found, data is set to zero\n"
               "-longheader  version of CLM file is set to 2\n"
               "-json        an additional JSON metafile with suffix '%s' is created\n"
               "grid_old.clm corresponding grid filename of input data,\n"
               "             can be omitted for metafile input\n"
               "grid_new_clm grid filename data should be regridded to\n"
               "data_old_clm filename of data that should be regridded\n"
               "data_new.clm filename of data file where regridded data is written\n\n"
               "(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file\n",
               progname,JSON_SUFFIX);
        return EXIT_SUCCESS;
      }
      else if(!strcmp(argv[1],"-v") || !strcmp(argv[1],"--version"))
      {
        puts(getversion());
        return EXIT_SUCCESS;
      }
      if(!strcmp(argv[iarg],"-size4"))
        isint=TRUE;
      else if(!strcmp(argv[iarg],"-byte"))
        isbyte=TRUE;
      else if(!strcmp(argv[iarg],"-metafile"))
        ismeta=TRUE;
      else if(!strcmp(argv[iarg],"-longheader"))
        setversion=2;
      else if(!strcmp(argv[iarg],"-search"))
        issearch=TRUE;
      else if(!strcmp(argv[iarg],"-json"))
        isjson=TRUE;
      else if(!strcmp(argv[iarg],"-zero"))
        iszero=TRUE;
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n"
                ERR_USAGE,argv[iarg],progname,progname);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<(ismeta ? 3 : 4)+iarg)
  {
    fprintf(stderr,"Error: Missing arguments.\n"
            ERR_USAGE,progname,progname);
    return EXIT_FAILURE;
  }
  format=CLM;
  if(ismeta)
  {
    index_datafile=(argc==iarg+3) ? iarg+1 : iarg+2;
    index_gridfile=(argc==iarg+3) ? iarg : iarg+1;
    header.cellsize_lon=header.cellsize_lat=0.5;
    header.firstyear=1901;
    header.firstcell=0;
    header.nyear=1;
    header.nbands=1;
    header.nstep=12;
    header.timestep=1;
    header.datatype=LPJ_SHORT;
    header.order=CELLYEAR;
    data_version=CLM_MAX_VERSION;
    data_file=openmetafile(&header,&metadata,&grid_name,NULL,&format,&swap,&offset,argv[index_datafile],TRUE);
    if(data_file==NULL)
      return EXIT_FAILURE;
    if(format==CLM)
    {
      if(freadheaderid(data_file,id,TRUE))
      {
        freemetadata(&metadata);
        free(grid_name.name);
        fclose(data_file);
        return EXIT_FAILURE;
      }
    }
    if(fseek(data_file,offset,SEEK_SET))
    {
      fprintf(stderr,"Error seeking in '%s' to offset %lu.\n",argv[index_datafile],offset);
      freemetadata(&metadata);
      free(grid_name.name);
      fclose(data_file);
      return EXIT_FAILURE;
    }
    if(argc==iarg+4)
    {
      free(grid_name.name);
      filename.name=strdup(argv[iarg]);
      if(filename.name==NULL)
      {
        printallocerr("name");
        freemetadata(&metadata);
        fclose(data_file);
        return EXIT_FAILURE;
      }
      filename.fmt=(setversion==2) ? CLM2 : CLM;
    }
    else
    {
      if(grid_name.name==NULL)
      {
        fprintf(stderr,"Error: grid filename must be specified in '%s' metafile.\n",argv[index_datafile]);
        freemetadata(&metadata);
        fclose(data_file);
        return EXIT_FAILURE;
      }
      filename.name=grid_name.name;
      filename.fmt=grid_name.fmt;
    }
  }
  else
  {
    index_datafile=iarg+2;
    index_gridfile=iarg+1;
    data_file=fopen(argv[iarg+2],"rb");
    if(data_file==NULL)
    {
      fprintf(stderr,"Error opening '%s': %s.\n",argv[iarg+2],strerror(errno));
      return EXIT_FAILURE;
    }
    data_version=setversion;
    if(freadanyheader(data_file,&header,&swap,id,&data_version,TRUE))
    {
      fprintf(stderr,"Error reading header in '%s'.\n",argv[iarg+2]);
      fclose(data_file);
      return EXIT_FAILURE;
    }
    if(data_version>CLM_MAX_VERSION)
    {
      fprintf(stderr,"Error: Unsupported version %d in '%s', must be less than %d.\n",
              data_version,argv[iarg+2],CLM_MAX_VERSION+1);
      fclose(data_file);
      return EXIT_FAILURE;
    }
    filename.name=strdup(argv[iarg]);
    if(filename.name==NULL)
    {
      printallocerr("name");
      fclose(data_file);
      return EXIT_FAILURE;
    }
    filename.fmt=(setversion==2) ? CLM2 : CLM;
  }
  grid=opencoord(&filename,TRUE);
  if(grid==NULL)
  {
    freemetadata(&metadata);
    free(filename.name);
    fclose(data_file);
    return EXIT_FAILURE;
  }
  ngrid=numcoord(grid);
  getcellsizecoord(&lon,&lat,grid);
  res.lon=lon;
  res.lat=lat;
  c=newvec(Coord,ngrid);
  if(c==NULL)
  {
    printallocerr("c");
    freemetadata(&metadata);
    free(filename.name);
    fclose(data_file);
    closecoord(grid);
    return EXIT_FAILURE;
  }
  for(i=0;i<ngrid;i++)
  {
    if(readcoord(grid,c+i,&res))
    {
      fprintf(stderr,"Error reading cell %d in '%s'.\n",i,filename.name);
      free(c);
      freemetadata(&metadata);
      free(filename.name);
      fclose(data_file);
      closecoord(grid);
      return EXIT_FAILURE;
    }
    //printf("c:%g %g\n",c[i].lon,c[i].lat);
  }
  closecoord(grid);
  free(filename.name);
  filename.name=strdup(argv[index_gridfile]);
  if(filename.name==NULL)
  {
    printallocerr("name");
    freemetadata(&metadata);
    fclose(data_file);
    return EXIT_FAILURE;
  }
  filename.fmt=(setversion==2) ? CLM2 : CLM;
  grid=opencoord(&filename,TRUE);
  if(grid==NULL)
  {
    free(c);
    freemetadata(&metadata);
    free(filename.name);
    fclose(data_file);
    return EXIT_FAILURE;
  }
  grid_type=getcoordtype(grid);
  ngrid2=numcoord(grid);
  getcellsizecoord(&lon,&lat,grid);
  res2.lon=lon;
  res2.lat=lat;
  if(res.lon!=res2.lon)
    fprintf(stderr,"Warning: longitudinal resolution %g in '%s' differs from %g in '%s.\n",res2.lon,argv[index_gridfile],res.lon,argv[index_datafile]);
  if(res.lat!=res2.lat)
    fprintf(stderr,"Warning: latitudinal resolution %g in '%s' differs from %g in '%s.\n",res2.lat,argv[index_gridfile],res.lat,argv[index_datafile]);
  c2=newvec(Coord,ngrid2);
  if(c2==NULL)
  {
    free(c);
    freemetadata(&metadata);
    free(filename.name);
    fclose(data_file);
    closecoord(grid);
    printallocerr("c2");
    return EXIT_FAILURE;
  }
  for(i=0;i<ngrid2;i++)
  {
    if(readcoord(grid,c2+i,&res2))
    {
      free(c);
      free(c2);
      freemetadata(&metadata);
      free(filename.name);
      fclose(data_file);
      closecoord(grid);
      fprintf(stderr,"Error reading cell %d in '%s'.\n",i,argv[index_gridfile]);
      return EXIT_FAILURE;
    }
    //printf("c2:%g %g\n",c2[i].lon,c2[i].lat);
  }
  closecoord(grid);
  if(res.lon!=header.cellsize_lon)
    fprintf(stderr,"Warning: longitudinal resolution %g in '%s' differs from %g in '%s.\n",header.cellsize_lon,argv[index_datafile],res.lon,filename.name);
  if(res.lat!=header.cellsize_lat)
    fprintf(stderr,"Warning: latitudinal resolution %g in '%s' differs from %g in '%s.\n",header.cellsize_lat,argv[index_datafile],res.lat,filename.name);
  if(header.nyear<=0)
  {
    fprintf(stderr,"Invalid nyear=%d, set to one.\n",header.nyear);
    header.nyear=1;
  }
  if(header.ncell!=ngrid)
  {
    fprintf(stderr,"Invalid number of cells %d in '%s', not %d.\n",
            header.ncell,argv[index_datafile],ngrid);
    free(c);
    free(c2);
    freemetadata(&metadata);
    free(filename.name);
    fclose(data_file);
    return EXIT_FAILURE;
  }
  if(ismeta)
    size=getfilesizep(data_file)-offset;
  else
    size=getfilesizep(data_file)-headersize(id,data_version);
  if(data_version>=3)
  {
    isint=header.datatype==LPJ_INT || header.datatype==LPJ_FLOAT;
    isbyte=header.datatype==LPJ_BYTE;
  }
  if(iszero)
  {
    zero=newvec(int,header.nbands*header.nstep);
    if(zero==NULL)
    {
      printallocerr("zero");
      free(c);
      free(c2);
      freemetadata(&metadata);
      free(filename.name);
      fclose(data_file);
      return EXIT_FAILURE;
    }
    if(header.datatype==LPJ_FLOAT)
    {
      fzero=(float *)zero;
      for(i=0;i<header.nbands*header.nstep;i++)
        fzero[i]=0;
    }
    else
      for(i=0;i<header.nbands*header.nstep;i++)
        zero[i]=0;
  }
  if(isint)
  {
    idata=newvec(int,(long long)header.ncell*header.nbands*header.nstep);
    if(idata==NULL)
    {
      printallocerr("idata");
      free(zero);
      free(c);
      free(c2);
      freemetadata(&metadata);
      free(filename.name);
      fclose(data_file);
      return EXIT_FAILURE;
    }
    if(size!=(long long)header.ncell*header.nyear*header.nbands*header.nstep*sizeof(int))
    {
      if((long long)header.ncell*header.nyear*header.nbands*header.nstep*sizeof(short)==size)
      {
        fprintf(stderr,"File size of '%s' does not match header, set datatype to 2 byte size.\n",argv[index_datafile]);
        isint=FALSE;
        free(idata);
        idata=NULL;
        data=newvec(short,(long long)header.ncell*header.nbands*header.nstep);
        if(data==NULL)
        {
          printallocerr("data");
          free(zero);
          free(c);
          free(c2);
          freemetadata(&metadata);
          free(filename.name);
          fclose(data_file);
          return EXIT_FAILURE;
        }
      }
      else
      {
        header.nyear=size/(sizeof(int)*header.ncell*header.nbands*header.nstep);
        if(size % (sizeof(int)*header.ncell*header.nbands*header.nstep)==0)
          fprintf(stderr,"File size of '%s' does not match header, number of years set to %d.\n",argv[index_datafile],header.nyear);
        else
        {
          fprintf(stderr,"File size of '%s' is not multiple of ncell and nbands.\n",argv[index_datafile]);
          free(idata);
          free(zero);
          free(c);
          free(c2);
          freemetadata(&metadata);
          free(filename.name);
          fclose(data_file);
          return EXIT_FAILURE;
        }
      }
    }
  }
  else if(isbyte)
  {
    bdata=newvec(Byte,(long long)header.ncell*header.nbands*header.nstep);
    if(bdata==NULL)
    {
      printallocerr("bdata");
      free(zero);
      free(c);
      free(c2);
      freemetadata(&metadata);
      free(filename.name);
      fclose(data_file);
      return EXIT_FAILURE;
    }
    if(size!=(long long)header.ncell*header.nyear*header.nbands*header.nstep)
    {
      header.nyear=size/(header.ncell*header.nbands*header.nstep);
      if(size % (header.ncell*header.nbands*header.nstep)==0)
        fprintf(stderr,"File size of '%s' does not match header, number of years set to %d.\n",argv[index_datafile],header.nyear);
      else
      {
        fprintf(stderr,"File size of '%s' is not multiple of ncell and nbands.\n",argv[index_datafile]);
        free(zero);
        free(bdata);
        free(c);
        free(c2);
        freemetadata(&metadata);
        free(filename.name);
        fclose(data_file);
        return EXIT_FAILURE;
      }
    }
  }
  else
  {
    data=newvec(short,(long long)header.ncell*header.nbands*header.nstep);
    if(data==NULL)
    {
      printallocerr("data");
      free(zero);
      free(c);
      free(c2);
      freemetadata(&metadata);
      free(filename.name);
      fclose(data_file);
      return EXIT_FAILURE;
    }
    if(size!=(long long)header.ncell*header.nyear*header.nbands*header.nstep*sizeof(short))
    {
      if((long long)header.ncell*header.nyear*header.nbands*header.nstep*sizeof(int)==size)
      {
        fprintf(stderr,"File size of '%s' does not match header, set datatype to 4 byte size.\n",argv[index_datafile]);
        isint=TRUE;
        free(data);
        data=NULL;
        idata=newvec(int,(long long)header.ncell*header.nbands*header.nstep);
        if(idata==NULL)
        {
          printallocerr("idata");
          free(zero);
          free(c);
          free(c2);
          freemetadata(&metadata);
          free(filename.name);
          fclose(data_file);
          return EXIT_FAILURE;
        }
      }
      else
      {
        header.nyear=size/(sizeof(short)*header.ncell*header.nbands*header.nstep);
        if(size % (sizeof(short)*header.ncell*header.nbands*header.nstep)==0)
          fprintf(stderr,"File size of '%s' does not match header, number of years set to %d.\n",argv[index_datafile],header.nyear);
        else
        {
          fprintf(stderr,"File size of '%s' is not multiple of ncell and nbands.\n",argv[index_datafile]);
          free(zero);
          free(data);
          free(c);
          free(c2);
          freemetadata(&metadata);
          free(filename.name);
          fclose(data_file);
          printallocerr("idata");
          return EXIT_FAILURE;
        }
      }
    }
  }
  file=fopen(argv[index_datafile+1],"wb");
  if(file==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s.\n",argv[index_datafile+1],strerror(errno));
    free(zero);
    free(data);
    free(bdata);
    free(idata);
    free(c);
    free(c2);
    freemetadata(&metadata);
    free(filename.name);
    fclose(data_file);
    printallocerr("idata");
    return EXIT_FAILURE;
  }
  header2=header;
  header2.cellsize_lon=(float)res2.lon;
  header2.cellsize_lat=(float)res2.lat;
  header2.ncell=ngrid2;
  if(format==CLM)
  {
    if(fwriteheader(file,&header2,id,data_version))
    {
      fprintf(stderr,"Error writing header in '%s'.\n",argv[index_datafile+1]);
      free(zero);
      free(data);
      free(bdata);
      free(idata);
      free(c);
      free(c2);
      freemetadata(&metadata);
      free(filename.name);
      fclose(data_file);
      fclose(file);
      printallocerr("idata");
      return EXIT_FAILURE;
    }
  }
  index=newvec(int,ngrid2);
  if(index==NULL)
  {
    printallocerr("index");
    free(zero);
    free(data);
    free(bdata);
    free(idata);
    free(c);
    free(c2);
    freemetadata(&metadata);
    free(filename.name);
    fclose(data_file);
    fclose(file);
    return EXIT_FAILURE;
  }
  for(i=0;i<ngrid2;i++)
  {
    j=findcoord(c2+i,c,&res,ngrid);
    if(j==NOT_FOUND) /* coordinate not found? */
    {
      if(issearch)
      {
        fputs("Coordinate ",stderr);
        fprintcoord(stderr,c2+i);
        fputs(" not found, replaced by ",stderr);
        index[i]=findnextcoord(&dist_min,c2+i,c,ngrid);
        fprintcoord(stderr,c+index[i]);
        fprintf(stderr,", distance=%g\n",dist_min);
      }
      else
      {
        fputs("Coordinate ",stderr);
        fprintcoord(stderr,c2+i);
        fprintf(stderr," at index %d not found.\n",i);
        index[i]=-1;
        if(!iszero)
        {
          free(zero);
          free(data);
          free(bdata);
          free(idata);
          free(index);
          free(c);
          free(c2);
          freemetadata(&metadata);
          free(filename.name);
          fclose(data_file);
          fclose(file);
          return EXIT_FAILURE;
        }
      }
    }
    else
      index[i]=j;
  }
  if(isint)
    for(i=0;i<header2.nyear;i++)
    {
      if(freadint(idata,(long long)header.ncell*header.nbands*header.nstep,swap,data_file)!=(long long)header.ncell*header.nbands*header.nstep)
      {
        fprintf(stderr,"Error reading '%s' in year %d.\n",argv[index_datafile],header2.firstyear+i);
        free(zero);
        free(idata);
        free(index);
        free(c);
        free(c2);
        freemetadata(&metadata);
        free(filename.name);
        fclose(data_file);
        fclose(file);
        return EXIT_FAILURE;
      }
      for(j=0;j<header2.ncell;j++)
      {
        if(index[j]==-1)
          fwrite(zero,sizeof(int),header.nbands*header.nstep,file);
        else
        {
          if(fwrite(idata+index[j]*header.nbands*header.nstep,sizeof(int),header.nbands*header.nstep,file)!=header.nbands*header.nstep)
            fprintf(stderr,"Error writing file '%s: %s.\n",argv[index_datafile+1],strerror(errno));
        }
      }
    }
  else if(isbyte)
    for(i=0;i<header2.nyear;i++)
    {
      if(fread(bdata,1,(long long)header.ncell*header.nbands*header.nstep,data_file)!=(long long)header.ncell*header.nbands*header.nstep)
      {
        fprintf(stderr,"Error reading '%s' in year %d.\n",argv[index_datafile],header2.firstyear+i);
        free(zero);
        free(bdata);
        free(index);
        free(c);
        free(c2);
        freemetadata(&metadata);
        free(filename.name);
        fclose(data_file);
        fclose(file);
        return EXIT_FAILURE;
      }
      for(j=0;j<header2.ncell;j++)
      {
        if(index[j]==-1)
          fwrite(zero,1,header.nbands*header.nstep,file);
        else
        {
          if(fwrite(bdata+index[j]*header.nbands*header.nstep,1,header.nbands*header.nstep,file)!=header.nbands*header.nstep)
            fprintf(stderr,"Error writing file '%s: %s.\n",argv[index_datafile+1],strerror(errno));
        }
      }
    }
  else
    for(i=0;i<header2.nyear;i++)
    {
      if(freadshort(data,(long long)header.ncell*header.nbands*header.nstep,swap,data_file)!=(long long)header.ncell*header.nbands*header.nstep)
      {
        fprintf(stderr,"Error reading '%s' in year %d.\n",argv[index_datafile],header2.firstyear+i);
        free(index);
        free(zero);
        free(data);
        free(c);
        free(c2);
        freemetadata(&metadata);
        free(filename.name);
        fclose(data_file);
        fclose(file);
        return EXIT_FAILURE;
      }
      for(j=0;j<header2.ncell;j++)
      {
        if(index[j]==-1)
          fwrite(zero,sizeof(short),header.nbands*header.nstep,file);
        else
        {
          if(fwrite(data+index[j]*header.nbands*header.nstep,sizeof(short),header.nbands*header.nstep,file)!=header.nbands*header.nstep)
            fprintf(stderr,"Error writing file '%s: %s.\n",argv[index_datafile+1],strerror(errno));
        }
      }
    }
  free(index);
  free(bdata);
  free(idata);
  free(zero);
  free(data);
  free(c);
  free(c2);
  fclose(file);
  fclose(data_file);
  if(ismeta || isjson)
  {
    out_json=malloc(strlen(argv[index_datafile+1])+strlen(JSON_SUFFIX)+1);
    if(out_json==NULL)
    {
      printallocerr("filename");
      free(filename.name);
      freemetadata(&metadata);
      return EXIT_FAILURE;
    }
    strcat(strcpy(out_json,argv[index_datafile+1]),JSON_SUFFIX);
    arglist=catstrvec(argv,argc);
    if(arglist==NULL)
    {
      printallocerr("arglist");
      free(out_json);
      free(filename.name);
      freemetadata(&metadata);
      return EXIT_FAILURE;
    }
    file=fopen(out_json,"w");
    if(file==NULL)
    {
      printfcreateerr(out_json);
      free(out_json);
      free(arglist);
      free(filename.name);
      freemetadata(&metadata);
      return EXIT_FAILURE;
    }
    if(data_version<4)
      header2.nbands/=header2.nstep;
    fprintjson(file,argv[index_datafile+1],NULL,arglist,&header2,&metadata,&filename,grid_type,format,id,FALSE,data_version);
    free(out_json);
    free(arglist);
    fclose(file);
  }
  free(filename.name);
  freemetadata(&metadata);
  return EXIT_SUCCESS;
} /* of 'main' */
